aibolit/patterns/partially_synchronized_methods/partially_synchronized_methods.py

```diff
@@ -0,0 +1,53 @@
+# The MIT License (MIT)
+#
+# Copyright (c) 2020 Aibolit
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included
+# in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+from typing import List
+
+from aibolit.ast_framework import AST, ASTNodeType
+from aibolit.utils.ast_builder import build_ast
+
+
+class PartiallySynchronizedMethods:
+    """
+    Methods and constructors must have no 'synchronized' statements at all,
+    including statements in lambda functions and anonymous class methods,
+    or consist of single 'synchronized' statement (fully synchronized)
+    """
+
+    def value(self, filename: str) -> List[int]:
+        ast = AST.build_from_javalang(build_ast(filename))
+        lines: List[int] = []
+        for method_ast in ast.get_subtrees(
+            ASTNodeType.METHOD_DECLARATION, ASTNodeType.CONSTRUCTOR_DECLARATION
+        ):
+            method_declaration = method_ast.get_root()
+            is_fully_sync_method = len(method_declaration.body) == 1 and \
```

Comment:
> len(method_declaration.body) == 1

method declaration is a list and it may have zero length?

Reply:
can it have length > 1 ?

Reply:
It appeared that yes. It may have zero length if method is abstract or native (implemented with c++). I also found some examples of code with just public methods with no body. I couldn't manage to figure out how is this possible, but we have examples of such code in integration tests.

To your second question, what do you mean? Can any method has several statements or fully synchronized method. If any method, then sure, it can have as many statements as it needs. But fully synchronized methods consist only of a single `synchronized` statement, with so,e statements nested in it. Notice, that `method_declaration.body` only contains top level statements, same as in `javalang`.

Reply:
If `method_declaration.body` is either None or some value, then I beg a question: Why we need `list` type for `body` property?

Reply:
There **were** some situations, where this field was `None`. I changed this behavior, so in such case we replace it with empty list. So it is always a list, but sometimes empty.

Reply:
Why we need `list` type to store optional value?

Reply:
@acheshkov  Because now you can iterate over the `body` nevertheless it is empty or not. If there is chances it is `None`, then you have to check it every time before `for` cycle. Now the value is consistent, and can always be iterated over.

Reply:
clear